src/test/csfle.rs

```diff
@@ -0,0 +1,1184 @@
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn views_prohibited() -> Result<()> {
+    if !check_env("views_prohibited", false) {
+        return Ok(());
+    }
+    let _guard = LOCK.run_exclusively().await;
+
+    // Setup: db initialization.
+    let (client, _) = init_client().await?;
+    client
+        .database("db")
+        .collection::<Document>("view")
+        .drop(None)
+        .await?;
+    client
+        .database("db")
+        .run_command(doc! { "create": "view", "viewOn": "coll" }, None)
+        .await?;
+
+    // Setup: encrypted client.
+    let auto_enc_opts = AutoEncryptionOptions::builder()
+        .key_vault_namespace(KV_NAMESPACE.clone())
+        .kms_providers(LOCAL_KMS.clone())
+        .extra_options(EXTRA_OPTIONS.clone())
+        .build();
+    let client_encrypted =
+        Client::with_encryption_options(CLIENT_OPTIONS.get().await.clone(), auto_enc_opts).await?;
+
+    // Test: auto encryption fails on a view
+    let result = client_encrypted
+        .database("db")
+        .collection::<Document>("view")
+        .insert_one(doc! {}, None)
+        .await;
+    let err = result.unwrap_err();
+    assert!(
+        err.to_string().contains("cannot auto encrypt a view"),
+        "unexpected error: {}",
+        err
+    );
+
+    Ok(())
+}
+
+macro_rules! failure {
+    ($($arg:tt)*) => {{
+        Box::new(crate::error::Error::internal(format!($($arg)*))) as Box<dyn std::error::Error>
+    }}
+}
+
+// TODO(RUST-36): use the full corpus with decimal128.
+fn load_corpus_nodecimal128(name: &str) -> Result<Document> {
+    let json: serde_json::Value = serde_json::from_str(&load_testdata_raw(name)?)?;
+    let mut new_obj = serde_json::Map::new();
+    let decimal = serde_json::Value::String("decimal".to_string());
+    for (name, value) in json.as_object().expect("expected object") {
+        if value["type"] == decimal {
+            continue;
+        }
+        new_obj.insert(name.clone(), value.clone());
+    }
+    let bson: bson::Bson = serde_json::Value::Object(new_obj).try_into()?;
+    match bson {
+        bson::Bson::Document(d) => Ok(d),
+        _ => Err(failure!("expected document, got {:?}", bson)),
+    }
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn corpus_coll_schema() -> Result<()> {
+    if !check_env("corpus_coll_schema", true) {
+        return Ok(());
+    }
+    let _guard = LOCK.run_exclusively().await;
+    run_corpus_test(false).await?;
+    Ok(())
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn corpus_local_schema() -> Result<()> {
+    if !check_env("corpus_local_schema", true) {
+        return Ok(());
+    }
+    let _guard = LOCK.run_exclusively().await;
+    run_corpus_test(true).await?;
+    Ok(())
+}
+
+async fn run_corpus_test(local_schema: bool) -> Result<()> {
```

Comment:
I'm curious if it would be any easier/more readable to model the corpus docs here with structs and use serde rather than parse them out directly. Was this something you tried/considered? There might be a lot of deserialization customization that would make it overkill...